### PR TITLE
Extract "nearby" constants (i+1,i-1), generate more instances of small numbers

### PR DESCRIPTION
**Pull request by agroce**
This PR handles (without memory) a number of examples echidna previously had trouble with, e.g:

```
contract NeedSmall {
  uint256 [] values;

  function push_it(uint256 v) public {
    values.push(v);
  }

  function echidna_nodups() public returns (bool) {
    uint256 length = values.length;
    uint256 i;
    for (i = 0; i < length; i++) {
      if (values[i] > 2) {
        if ((values[i] * values[i]) < 1048576) {
          return false;
        }
      }
    }
    return true;
  }
}
```

```
contract NeedOffByOne {
  uint256 [] values;

  function push_it(uint256 v) public {
    values.push(v);
  }

  function echidna_nodups() public returns (bool) {
    uint256 length = values.length;
    uint256 i;
    for (i = 0; i < length; i++) {
      if (values[i] > 81789178917) {
	if (values[i] < 81789178919) {
          return false;
	}
      }
    }
    return true;
  }
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# echidna
+# echidna
```
Comment (japesinator) [inline README.md]: Extra space

File: lib/Echidna/ABI.hs [hunk 2]
```diff
@@ -110,9 +110,13 @@ mkGenDict p vs cs = GenDict p (hashMapBy abiValueType vs) (hashMapBy (fmap $ fma
 
 -- Generation (synthesis)
 
+getRandomUint :: MonadRandom m => Int -> m Integer
+getRandomUint n =  getRandomR (1 :: Integer, 100 :: Integer) >>=
```
Comment (japesinator) [inline lib/Echidna/ABI.hs]: `fromList [(1, getRandomR (0, 1024), (9, getRandomR (0, 2 ^ n - 1)]`

Comment (japesinator) [reply lib/Echidna/ABI.hs]: Also, 1024 isn't a valid uint8

File: lib/Echidna/Solidity.hs [hunk 3]
```diff
@@ -174,7 +174,7 @@ extractConstants = nub . concatMap (constants "" . view contractAst) where
   -- 2.2: We're looking at something of the form @type: int_const [...]@, an integer literal
   --      @type: "int_const 123"@ ==> @[AbiUInt 8 123, AbiUInt 16 123, ... AbiInt 256 123]@
   constants "type"  (literal "int_const"      (as decimal) -> Just i) =
-    let l f = f <$> [8,16..256] <*> [fromIntegral (i :: Integer)] in l AbiInt ++ l AbiUInt
+    let l f = f <$> [8,16..256] <*> (fromIntegral <$> ([i, i+1, i-1] :: [Integer])) in l AbiInt ++ l AbiUInt
```
Comment (japesinator) [inline lib/Echidna/Solidity.hs]: `[i-1..i+1]` (or maybe i-8..i+8? idk)

File: lib/Echidna/Transaction.hs [hunk 4]
```diff
@@ -115,7 +115,7 @@ liftSH = S.state . runState . zoom hasLens
 -- 'Transaction's \"on-chain\".
 setupTx :: (MonadState x m, Has VM x) => Tx -> m ()
 setupTx (Tx c s r v) = S.state . runState . zoom hasLens . sequence_ $
-  [ result .= Nothing, state . pc .= 0, state . stack .= mempty, state . gas .= 0xffffffff
+  [ result .= Nothing, state . pc .= 0, state . stack .= mempty, state . memory .= mempty, state . gas .= 0xffffffff
```
Comment (japesinator) [inline lib/Echidna/Transaction.hs]: is this supposed to be in this PR? also the associated test

Comment (agroce) [reply lib/Echidna/Transaction.hs]: Hmm, not sure how this got in, yeah, it's a fix I was testing but not relevant here

**Comment by agroce** [comment]
I'll get on these tomorrow.  I think off-by-one handles most of the cases we want, and -8..+8 is too much.  I do want to do some light reverse engineering of simple constraints, though, e.g. EXPR * 4 < 8 mines 3/4/5 and 7/8/9 but also 1, 2, 3 (from computing 8 / 4).

But not in this PR

**Comment by agroce** [comment]
will finish the rest tomorrow...

**Comment by agroce** [comment]
No clue why these tests fail.

File: lib/Echidna/ABI.hs (hunk 2 above)
Comment (agroce) [reply lib/Echidna/ABI.hs]: Done

File: lib/Echidna/Solidity.hs (hunk 3 above)
Comment (agroce) [reply lib/Echidna/Solidity.hs]: Done

File: lib/Echidna/Transaction.hs (hunk 4 above)
Comment (agroce) [reply lib/Echidna/Transaction.hs]: Done

**Pull request merged**